src/envs/tools.py

```diff
@@ -0,0 +1,646 @@
+            wrench_sizes = [rng.uniform(0, 1) for _ in range(self.num_wrenches)]
+            sizes = screwdriver_sizes + hammer_sizes + wrench_sizes
+            for j, size in enumerate(sizes):
+                while True:
+                    pose_x = rng.uniform(self.table_lx, self.table_ux)
+                    pose_y = rng.uniform(self.table_ly, self.table_uy)
+                    # If collision with any contraption, try again
+                    if any(data[c][0] < pose_x < data[c][2] and \
+                           data[c][1] < pose_y < data[c][3]
+                           for c in contraptions):
+                        continue
+                    # If collision with any item or tool, try again
+                    if any(abs(data[it][0] - pose_x) < self.close_thresh and \
+                           abs(data[it][1] - pose_y) < self.close_thresh
+                           for it in items + tools):
+                        continue
+                    # Otherwise, we found a valid pose
+                    break
+                is_held = 0.0  # always start off not held
+                if j < self.num_screwdrivers:
+                    tool = Object(f"screwdriver{j}", self._screwdriver_type)
+                    shape = rng.uniform(0, 1)
+                    feats = [pose_x, pose_y, shape, size, is_held]
+                elif j < self.num_screwdrivers + self.num_hammers:
+                    ind = j - self.num_screwdrivers
+                    tool = Object(f"hammer{ind}", self._hammer_type)
+                    feats = [pose_x, pose_y, size, is_held]
+                else:
+                    ind = j - self.num_screwdrivers - self.num_hammers
+                    tool = Object(f"wrench{ind}", self._wrench_type)
+                    feats = [pose_x, pose_y, size, is_held]
+                tools.append(tool)
+                data[tool] = np.array(feats, dtype=np.float32)
+            state = State(data)
+            tasks.append(Task(state, goal))
+        return tasks
+
+    @staticmethod
+    def _HandEmpty_holds(state: State, objects: Sequence[Object]) -> bool:
+        robot, = objects
+        return state.get(robot, "fingers") > 0.5
+
+    @staticmethod
+    def _Holding_holds(state: State, objects: Sequence[Object]) -> bool:
+        # Works for any item or tool
+        item_or_tool, = objects
+        return state.get(item_or_tool, "is_held") > 0.5
+
+    @staticmethod
+    def _OnTable_holds(state: State, objects: Sequence[Object]) -> bool:
+        # Works for any item
+        item, = objects
+        return state.get(item, "on_table") > 0.5
+
+    def _Placed_holds(self, state: State, objects: Sequence[Object]) -> bool:
+        # Works for any item
+        item, contraption = objects
+        pose_x = state.get(item, "pose_x")
+        pose_y = state.get(item, "pose_y")
+        return self._is_pose_on_contraption(state, pose_x, pose_y, contraption)
```

Comment:
we should be able to learn this as two different predicates in our current grammar, right?

Reply:
4 because of the lower and upper bounds, but yeah theoretically... it's a goal predicate though so we don't have to worry about this